### ISSUE TITLE
Add an interface that lets other applications script fman

### ISSUE THREAD

**Issue by mherrmann** [fman-community/fman#76]
For example on the command line: `fman -c "show_alert('hello world')"` would show an alert inside fman. Likewise, such a solution could be used to query the state of fman. The respective results could be returned in JSON.

**Comment by raguay** [comment]
On macOS, it would be nice to have a full scripting dictionary. That way I could use JavaScript for Automation (JXA) or Swift in Scriptarian to perform tasks.

**Comment by mherrmann** [comment]
What do you mean by a scripting dictionary?

**Comment by raguay** [comment]
https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/ScriptableCocoaApplications/SApps_intro/SAppsIntro.html#//apple_ref/doc/uid/TP40001982-BCICHGIE
https://developer.apple.com/library/prerelease/content/documentation/AppleScript/Conceptual/AppleScriptX/Concepts/scriptable_apps.html

These articles tell what has to be done to add a scripting dictionary to an application. Once done, others can write scripts that interact with the application using AppleScript, JXA (JavaScript for Automation), Python using the Scripting bridge (I've seen it before, but I forgot the reference to it), Scriptarian (A Swift version of the Apple Script Editor), etc.

It adds more flexibility to the application and makes it more attractive to power users. 

The interface you describe above would be for every OS, but this would be extra just for macOS.

**Comment by mherrmann** [comment]
Thank you for the clarification!

**Comment by bmix** [comment]
The problem with native OS scripting is, that it is difficult to maintain for the developer of the application, that is to be scripted.

I usually tend to favor applications, that use the OS' native scripting capabilities (on Windows this is _ActiveScript_, or _Windows Script Host (WSH)_, as many call it), on Apple it is the _OSA (Open Scripting Architecture)_ subsystem (on KDE and Gnome some default scripting engine exists, as well, as far as I know), because these, being generic scripting hosts, allow us to use our language of choice, which tends to be ECMAScript. Others may want to use Python and others may want to use something else (see the Wikipedia entry on [WSH→Available Scripting Engines](https://en.wikipedia.org/wiki/Windows_Script_Host#Available_scripting_engines), which lists a little bit too many, I doubt ActiveScripting implementations exist for all of these, but I know for sure, that all the Rexx derivates, listed there are valid, Python, Lua, Ruby, Perl, PHP and some others are as well). 

On macOS (OSA scripting) one can use AppleScript, Javascript and maybe others.

I am not sure about the state of KDE scripting, but once it could be done via dbus messages. Gnome offers some scripting-tech, as well, IMO.

For the user, using the host's generic scripting engine (if such exists) may be the best option, because it allows easy access to the rest of the system, as well as the choice of programming-language. But for the developer, @mherrmann, it means implementing a different scripting host for each operating-system. 

I feel, that the ActiveScript Host is not a moving target anymore, it is well understood and I could point you to a basic reference implementation, that is open source, however, there is at least two ways to implement it, one means your application hosts all itself, isolating the scripts from the rest of the system, or you must deliver a COM interface, making your application one of many COMs installed on the system and thus shareable. But don't quote me on that. :^)

**Comment by mherrmann** [comment]
Thank you very much for your input Andreas. What you're saying sounds very reasonable. I agree that for the user, the host's scripting language is (probably) the preferred choice, especially because of interoperability with the OS's other features. But you are (unfortunately) also correct about this being the most difficult to implement.

What if instead, fman opened a port on localhost that you could send commands to? That would be trivial to implement. It could be made configurable so the port isn't opened unless the user configures it to be opened. There are also some security considerations, but I think they're manageable. How good are the various scripting languages at sending HTTP requests?

**Comment by brodo** [comment]
Why not just open a local TCP port and listen for messages separated by newlines? This would be even easier than the HTTP server and one could write batch scripts using netcat very easily.

**Comment by mherrmann** [comment]
Yes, something along those lines! Doesn't have to be HTTP.

**Comment by bmix** [comment]
I agree, this would be nice. Also I think a simple local TCP port would suffice. That could be used from shell-scripts, as well. However, there would arise the need for return values, so it'd be two-way communication. If you'd do it as httpd, than it should be trivial to limit accepted connections to a range of IPs, set by the user (usually localhost or link-local).

But whatever you do, you'd need to define a language agnostic API, which then would be used by the Python scripting backend as well, just by other means (direct, not over TCP).

**Comment by mherrmann** [comment]
I think my preference would be HTTP with JSON as a means of communication. A language agnostic API which is used by the internal Python system would of course be the apex of elegance. But I don't know if it will be feasible from a performance perspective.

**Comment by raguay** [comment]
If you open ports, you need to think about security. It would be best to have a way to specify what networks systems can send requests from. For most, the local network netmask would do, but you would need a high security option for when traveling and on an open network. When detecting a high secure setting, only accepting requests from the localhost connect would be best.

Once you have something like this in place, you can create small applications that run as a scripting bridge between the Oses scripting system and fman. Then users could use the scripting bridge of their choice. It might be a little more flexible.

It would need a way to call extensions as well. Not just internal stuff. 😄

**Comment by mherrmann** [comment]
Yes, I would only allow connections from localhost. Plus there'd need to be some authentication. Some JetBrains IDEs had a built-in server that was running on localhost and could be exploited by JavaScript running on arbitrary web sites.

Yes you're right, this could act as one end of a scripting bridge :)

I guess, ideally, it would give you access to the full plugin API. That API already includes `DirectoryPane#run_command` and `run_application_command` for running plugins.
